src/state_flow/state.clj

```
@@ -33,9 +36,14 @@
       (error-catching-state
        (fn [s]
          (let [[v s'] ((p/-extract fv) s)]
-           (if (e/failure? v)
-             [v s']
-             [(result-or-err f v) s'])))))
+           (cond (e/failure? v)
```

I don't think this is the right abstraction level to implement this change.
We should be storing assertion results in the meta of the state (if not, we should be doing that with the new `match?`)
We can choose to throw/stop if the assertion being stored is a failed assertion or something like that.
This will result in an incomplete test report. Maybe we should be aiming to stop outputting failures instead of stopping the test.

---

If you store this option of the state meta, `match?` can access fetch the option from the meta and throw after a test fails. Throwing will shortcut the execution. I think this is simple and effective. WDYT?

---

I didn't want achieve shortcircuiting via raising an exception because then we would be mixing that special behavior with the normal exception handling flow. Doing so would mean that when shortcircuiting is enabled a test failure would also appear as an exception, right? Does this seem like a valid thing to be concerned about mixing?

The idea of continuing to run the tests but only reporting the first failure feels a little strange to me. Like what if the subsequent steps are kinda slow? Or what if you want to leave the system at the exact state that it was at after the first assertion failure so that you can inspect it?

That said, storing the flag in the meta of the state sounds cleaner than using `binding`. I can look to see if that comes together well

---

Technically, the state monad implementation will short circuit on `failure?` as defined here: https://github.com/nubank/cats/blob/master/src/cats/monad/exception.cljc#L151

So having `match?` return a failure when fail-fast is turned on should also short circuit the run.
However, the error-handler probably needs to distinguish between this and other exceptions if we don't want to just throw it.

---

Ah, right; I had considered that in the beginning but decided it would be better to avoid encoding a "special exception" that we treat differently. In the end my implementation does the exact same things with a "special return value" that stops further execution. So I guess it comes down to if we want to hijack the exceptions control flow or the normal control flow. I suppose since exceptions always stop execution, using that flow is a little more in-line, and then the only exceptional case we encode is whether we throw the exception or swallow it (for test assertion failures) in the end.

So, all that said, I should switch the implementation to use a special exception and toggle the short-circuit behavior via a meta flag in the state?

---

That looks good to me :)
Just to clarify, I don't think you need to throw an exception, just return a failure data structure should be equivalent. I don't think I ever tested this though :thinking: 

Let me know if you want help with that.